msal/wam.py

```diff
@@ -0,0 +1,159 @@
+"""This module acquires token via WAM, with the help of Mid-tier.
+
+Mid-tier supports these Windows versions
+https://github.com/AzureAD/microsoft-authentication-library-for-cpp/pull/2406/files
+"""
+from threading import Event
+import json
+import logging
+
+import pymsalruntime  # See https://github.com/AzureAD/microsoft-authentication-library-for-cpp/pull/2419/files#diff-d5ea5122ff04e14411a4f695895c923daba73c117d6c8ceb19c4fa3520c3c08a
+import win32gui  # Came from package pywin32
+
+logger = logging.getLogger(__name__)
+
+
+class _CallbackData:
+    def __init__(self):
+        self.signal = Event()
+        self.auth_result = None
+
+    def complete(self, auth_result):
+        self.signal.set()
+        self.auth_result = auth_result
+
+
+def _read_account_by_id(account_id):
+    callback_data = _CallbackData()
+    pymsalruntime.read_account_by_id(
+        account_id,
+        "correlation_id",
+        lambda result, callback_data=callback_data: callback_data.complete(result)
+        )
+    callback_data.signal.wait()
+    return callback_data.auth_result
+
+
+def _convert_result(result):  # Mimic an on-the-wire response from AAD
+    error = result.get_error()
+    if error:
+        return {
+            "error": "broker_error",
+            "error_description": "{}. Status: {}, Error code: {}, Tag: {}".format(
+                error.get_context(),  # Available since pymsalruntime 0.0.4
+                error.get_status(), error.get_error_code(), error.get_tag()),
+            }
+    id_token_claims = json.loads(result.get_id_token()) if result.get_id_token() else {}
+    account = result.get_account()
+    assert account.get_account_id() == id_token_claims.get("oid"), "Emperical observation"  # TBD
+    return {k: v for k, v in {
+        "access_token": result.get_access_token(),
+        "expires_in": result.get_access_token_expiry_time(),
+        #"scope": result.get_granted_scopes(),  # TODO
```

Comment:
>result.get_granted_scopes() [](http://example.com/codeflow?start=18&length=27)

@rayluo , in MSAL CPP, this is a map, what is the string format you want me to return?

Reply:
Interesting. ~Why is it a map?~ Anyway. Conceptually, the scopes defined in OAuth2 RFC is a list of strings, but can be represented as one long space-delimited string.

In our case, my understanding is mid-tier data types do not include list. So, let's go with the space-delimited string, for example `"scope1 scope2 scope3"`. (Note that, the scopes are supposed to be orderless, but it does not harm if mid-tier would return a pre-sorted scope list, represented as space-delimited string.)